### PR TITLE
[Keyboard] Add Spleeb default and custom firmware

### PR DESCRIPTION
**Pull request by chrishoage**
## Description

This adds the keyboard configuration and a custom keymap. 

I did my best to follow the PR checklist, notably everything custom in terms of oled, encoder mode map, and custom transport is not enabled in the `default` keymap. 

Happy to adjust anything as needed.  

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/spleeb/halconf.h [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// Copyright 2022 Chris Hoage (@chrishoage)
```
Comment (waffle87) [inline keyboards/spleeb/halconf.h:1]: This file can be removed (`BOARD = QMK_PM2040` enables I2C)

File: keyboards/spleeb/info.json [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        "rotary": [
+            {
+                "pin_a": "GP4",
+                "pin_b": "GP21",
+                "resolution": 2
+            }
+        ]
+    },
+    "features": {
+        "encoder": true,
```
Comment (waffle87) [inline keyboards/spleeb/info.json:18]:
```suggestion
```

File: keyboards/spleeb/info.json [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        "nkro": true
+    },
+    "matrix_pins": {
+        "cols": ["GP23", "GP20", "GP22", "GP26", "GP27", "GP28", "GP29"],
+        "rows": ["GP5", "GP6", "GP7", "GP8", "GP9"]
+    },
+    "matrix_size": {
+        "cols": 7,
+        "rows": 10
+    },
```
Comment (waffle87) [inline keyboards/spleeb/info.json:28]:
```suggestion
```

File: keyboards/spleeb/readme.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
+# spleeb
+
+![spleeb](https://github.com/chrishoage/spleeb/raw/main/images/built_spleeb_1.jpg)
```
Comment (waffle87) [inline keyboards/spleeb/readme.md:3]: This image should be rehosted on Imgur.

**Review by waffle87 — CHANGES_REQUESTED** [review]

**Comment by chrishoage** [comment]
@waffle87 thank you very much for your review! I have applied your suggestions and force pushed over the previous commit.

File: keyboards/spleeb/readme.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
+# spleeb
+
+![spleeb](https://i.imgur.com/2rmZa6M.jpg)
```
Comment (drashna) [inline keyboards/spleeb/readme.md:3]: This uses a smaller, but still high res version of the image that is more comaptible with github's markdown rendering. 
```suggestion
![spleeb](https://i.imgur.com/2rmZa6Mh.jpg)
```

**Review by drashna — COMMENTED** [review]
Just a heads up: https://github.com/qmk/qmk_firmware/pull/18218

Nothing needs to change here, but some of the kb code may/will be made redundant by that PR.

File: keyboards/spleeb/readme.md (hunk 5 above)
Comment (chrishoage) [reply keyboards/spleeb/readme.md]: Thank you! I addressed this in the fixup commit.

**Comment by chrishoage** [comment]
> Just a heads up: #18218
> 
> Nothing needs to change here, but some of the kb code may/will be made redundant by that PR.

Yes I actually based my code on this MR. I left a note pointing to that PR 

https://github.com/qmk/qmk_firmware/blob/4c5069c593f8c1ffa8a77d13a6594edb77d5abe1/keyboards/spleeb/spleeb.c#L285-L290

However since this is targeting master I wanted to be able to use the same behavior now. I plan to refactor to use the new pointer modes when they are released to master in the future.

**Review by drashna — APPROVED** [review]

**Comment by chrishoage** [comment]
This PR is not stale, it is just waiting on a review. Please do not close.

File: keyboards/spleeb/mcuconf.h [hunk 6]
```diff
@@ -0,0 +1,9 @@
+// Copyright 2022 Chris Hoage (@chrishoage)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include_next "mcuconf.h"
```
Comment (fauxpark) [inline keyboards/spleeb/mcuconf.h:6]:
```suggestion
#include_next <mcuconf.h>
```

File: keyboards/spleeb/config.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
+// Copyright 2022 Chris Hoage (@chrishoage)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include "config_common.h"
+
```
Comment (fauxpark) [inline keyboards/spleeb/config.h:7]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by chrishoage** [comment]
@fauxpark Thank you! I have applied your suggestions.

File: keyboards/spleeb/spleeb.c [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+ * Includes:
+ *   - raw value
+ *   - drag-scroll: on/off
+ *   - sniping: on/off
+ *   - default DPI: internal table index/actual DPI
+ *   - sniping DPI: internal table index/actual DPI
+ */
+static void debug_spleeb_config_to_console(spleeb_config_t* config) {
+#    ifdef CONSOLE_ENABLE
+    dprintf("(spleeb) process_record_kb: config = {\n"
```
Comment (drashna) [inline keyboards/spleeb/spleeb.c:328]:
```suggestion
    pd_dprintf("(spleeb) process_record_kb: config = {\n"
```

File: keyboards/spleeb/spleeb.h [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    SNIPING_MODE_TOGGLE,
+    DRAGSCROLL_MODE,
+    DRAGSCROLL_MODE_TOGGLE,
+#endif // POINTING_DEVICE_ENABLE
+#ifdef SPLEEB_ENCODER_MODE_MAP_ENABLE
+    ENC_MODE_STEP_LEFT,
+    ENC_MODE_STEP_RIGHT,
+#endif // SPLEEB_ENCODER_MODE_MAP_ENABLE
+    SPLEEB_SAFE_RANGE,
+};
```
Comment (drashna) [inline keyboards/spleeb/spleeb.h:33]:
```suggestion
enum spleeb_keycodes {
    POINTER_DEFAULT_DPI_FORWARD = QK_KB,
    POINTER_DEFAULT_DPI_REVERSE,
    POINTER_SNIPING_DPI_FORWARD,
    POINTER_SNIPING_DPI_REVERSE,
    SNIPING_MODE,
    SNIPING_MODE_TOGGLE,
    DRAGSCROLL_MODE,
    DRAGSCROLL_MODE_TOGGLE,
    ENC_MODE_STEP_LEFT,
    ENC_MODE_STEP_RIGHT,
};
```

File: keyboards/spleeb/spleeb.h [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+#    define SP_RMOD POINTER_SNIPING_DPI_REVERSE
+#    define SNIPING SNIPING_MODE
+#    define SNP_TOG SNIPING_MODE_TOGGLE
+#    define DRGSCRL DRAGSCROLL_MODE
+#    define DRG_TOG DRAGSCROLL_MODE_TOGGLE
+#endif // POINTING_DEVICE_ENABLE
+#ifdef ENCODER_ENABLE
+#    define ENC_STL ENC_MODE_STEP_LEFT
+#    define ENC_STR ENC_MODE_STEP_RIGHT
+#endif // ENCODER_ENABLE
```
Comment (drashna) [inline keyboards/spleeb/spleeb.h:48]:
```suggestion
#define DF_MOD POINTER_DEFAULT_DPI_FORWARD
#define DF_RMOD POINTER_DEFAULT_DPI_REVERSE
#define SP_MOD POINTER_SNIPING_DPI_FORWARD
#define SP_RMOD POINTER_SNIPING_DPI_REVERSE
#define SNIPING SNIPING_MODE
#define SNP_TOG SNIPING_MODE_TOGGLE
#define DRGSCRL DRAGSCROLL_MODE
#define DRG_TOG DRAGSCROLL_MODE_TOGGLE
#define ENC_STL ENC_MODE_STEP_LEFT
#define ENC_STR ENC_MODE_STEP_RIGHT
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by chrishoage** [comment]
@drashna thank you for your review! I have applied all suggestions as fixup commits.

**Review by drashna — APPROVED** [review]

File: keyboards/spleeb/info.json [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    "features": {
+        "nkro": true
+    },
+    "matrix_pins": {
+        "cols": ["GP23", "GP20", "GP22", "GP26", "GP27", "GP28", "GP29"],
+        "rows": ["GP5", "GP6", "GP7", "GP8", "GP9"]
+    },
+    "processor": "RP2040",
+    "secure": {
```
Comment (waffle87) [inline keyboards/spleeb/info.json:25]:
```suggestion
    "board": "QMK_PM2040",
    "secure": {
```

File: keyboards/spleeb/rules.mk [hunk 12]
```diff
@@ -0,0 +1,4 @@
+# Board (built with Blok but any Pro Micro pinout RP2040 should work)
+BOARD = QMK_PM2040
+
```
Comment (waffle87) [inline keyboards/spleeb/rules.mk:3]:
```suggestion
```

**Review by waffle87 — APPROVED** [review]

**Pull request merged**